scripts/check-docs/src/main.rs

```diff
@@ -0,0 +1,227 @@
+
+    let additional_warnings = valid_anchors
+        .iter()
+        .filter(|valid_anchor| {
+            let anchor_used_in_a_pair = pairs.iter().any(|(_, anchor)| anchor == *valid_anchor);
+            !anchor_used_in_a_pair
+        })
+        .map(|unused_anchor| anyhow!("Anchor unused: {unused_anchor:?}!"))
+        .collect::<Vec<_>>();
+
+    (errors, additional_warnings)
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+struct Include {
+    anchor_name: String,
+    anchor_file: PathBuf,
+    include_file: PathBuf,
+    line_no: usize,
+}
+
+fn parse_includes(text_w_includes: String) -> anyhow::Result<Vec<Include>, Error> {
+    let apply_regex = |regex: Regex| {
+        text_w_includes
+            .lines()
+            .filter_map(|line| regex.captures(line))
+            .map(|capture| {
+                let include_file = PathBuf::from(&capture[1]).canonicalize()?;
+                let line_no = capture[2].parse()?;
+                let anchor_file = PathBuf::from(&capture[3]);
+                let anchor_name = capture[4].to_owned();
+
+                let the_path = include_file.parent().unwrap().join(anchor_file);
+
+                let anchor_file = the_path.canonicalize();
+                if anchor_file.is_err() {
+                    panic!(
+                        "{the_path:?} when canonicalized gives error {:?}",
+                        anchor_file.err().unwrap()
+                    )
+                }
+
+                let anchor_file = anchor_file?;
+
+                Ok(Include {
+                    anchor_name,
+                    anchor_file,
+                    include_file,
+                    line_no,
+                })
+            })
+            .collect::<Result<Vec<_>, Error>>()
+    };
+
+    apply_regex(Regex::new(
+        r"^(\S+):(\d+):\s*\{\{\s*#include\s*(\S+)\s*:\s*(\S+)\s*\}\}",
+    )?)
+}
+
+fn filter_valid_anchors(starts: Vec<Anchor>, ends: Vec<Anchor>) -> (Vec<Anchor>, Vec<Error>) {
+    let find_anchor_end_by_name = |anchor_name: &str, file: &Path| {
+        ends.iter()
+            .filter(|el| el.name == *anchor_name && el.file == file)
+            .collect::<Vec<_>>()
+    };
+
+    let (pairs, errors):(Vec<_>, Vec<_>) = starts.into_iter().map(|start| {
+        let matches_by_name = find_anchor_end_by_name(&start.name, &start.file);
+
+        let (begin, end) = match matches_by_name.as_slice() {
+            [single_match] => Ok((start, (*single_match).clone())),
+            [] => Err(anyhow!("Couldn't find a matching end anchor for {start:?}")),
+            multiple_ends => Err(anyhow!("Found too many matching anchor ends for anchor: {start:?}. The matching ends are: {multiple_ends:?}")),
+        }?;
+
+        match check_validity_of_anchor_pair(&begin, &end) {
+            None => Ok((begin, end)),
+            Some(_) => {
+                let err_msg = check_validity_of_anchor_pair(&begin, &end).iter().map(|e|e.to_string()).collect::<Vec<_>>().join("\n");
```

Comment:
You're calling the validating function twice. Perhaps just use the result ignored in `Some(_)`. That should be the returned error.